design/build-cache.md

```diff
@@ -0,0 +1,201 @@
+    cache: ["my-cache-1"]
+    requires: [~commit]
+    image: rust:latest
+    steps:
+      - add_musl: rustup target install x86_64-unknown-linux-musl
+      - build: cargo build --release --target=x86_64-unknown-linux-musl
+```
+
+### Design considerations
+
+1. Resource usage
+   1. Copying to and from a remote server for every job in a pipeline can lead to unnecessary network bandwidth usage and unnecessarily slow speeds.
+2. Integrity and consistency of cached data.
+3. User experience
+   1. Is it easily adaptable to new use cases?
+   2. Is it easily enabled and administered?
+   3. Is it easily configurable by end users once enabled?
+   4. Does it speed up build time? How much does it speed up build time for various use cases?
+   5. Is it consistent?
+   6. Is it reliable?
+4. Security
+   1. Is it possible to inject malicious code into pipelines via the cache mechanism? Malicious pull requests are a good example of this.
+
+### Security
+
+Injecting malicious code into a build via the caching mechanism is a real concern. This can lead to theft of tokens or credentials from job, or in even lead to malicious code being deployed to a production environment. 
+
+Pull requests, as the main vector for injection of malicious code, require additional security considerations. Pull requests therefore treat caches as readonly. 
+
+Caching of potentially sensitive data also needs to be handled. Data provided via the secrets field should not be cached. 
+
+### Cache invalidation
+
+- [Proposed] Changing environment variables or secrets should invalidate existing caches. Mass invalidation can be avoided by including a hash of variables and secrets with the cache metadata.
+- [Proposed] The user should also be able to invalidate caches via the UI.
+
+### Conflicts
+
+Multiple caches trying to restore the same file/files will result in an error. Caches by default do not overwrite existing files or directories. This behaviour can be configured by setting selecting a merging strategy.
+
+The following merging strategies are defined for dealing with file conflicts when extracting cached data on ta local build node.
+
+1. merge: Compares creation time of files and directories. Recursively merges files and directories from multiple caches. Selects the *newest* version of a file.
+2. [Proposed] preserve: Compares creation time of files and directories. Recursively merges files and directories from multiple caches. Selects the *oldest* version of a file.
+3. reject: Fails the build if there are any file conflicts.
+
+
+### Scoping
+
+Cache scoping levels can be expressed as `Job level < Event level < Pipeline level`. Job level scoping limits the cache to particular jobs, within a single event, within a single pipeline. Event level scoping limits the cache to a certain event within a single pipeline. Pipeline level scoping represents a globally scoped cache for the single pipeline, with the cache being available for all instances of the pipeline.
```

Comment:
For the job level scope, should it be a particular job within a single event or across different events? I think the later is probably a more common use case. For example I'm publishing an npm package in my component job. When I do npm install, I would like the dependencies to be in npm cache if I haven't touched my package.json file from my last commit (this logic is managed by npm itself)

Reply:
Could you elaborate on the use case of a Event scoped - Job level cache?

Reply:
> Could you elaborate on the use case of a Event scoped - Job level cache?

This is the existing definition, right? As is says in the proposal, _Job level scoping limits the cache to particular jobs, within a single event, within a single pipeline. Event level scoping limits the cache to a certain event within a single pipeline. _.

Use case for job scoped caches, as in the current proposal:
 - Caching of files/artifacts/data that are only used from certain jobs.  

Use case for event scoped caches, as in the current proposal:
- Caching of files/artifacts/data that should be available to all jobs within an event.

Reply:
I guess there might be some misunderstanding, but I don't see how a event-scoped + job-level cache could be useful? Cache is only useful if multiple builds use it. But `event-scoped + job-level` essentially means only one build will be able able to use it?

Reply:
The current proposal with an event and job scoped cache is based on the premise that you might have something you want to cache, but for whatever reason you do not want said cache to be extracted into certain jobs. Maybe the data is only being used in 2 jobs, but you have 20 jobs in total. In such a case it might make more sense performance wise to only use caching on the 2 relevant jobs, instead of all jobs.

I still think reworking scoping from the current proposed format to the following might be a good idea.

> However, scoping could be reworked to be like this:

 - Event scoped, only valid for a single event
   - Job level caching
   - Pipeline level caching
 - Statically scoped, valid across all executions of a pipeline
   - Job level caching
   - Pipeline level caching

This will allow for more flexibility and it also removes the ambiguity between event and job scoped caches.